nav2_motion_primitives/README.md

```diff
@@ -1,2 +1,115 @@
+using SomePrimitiveResult = std_msgs::msg::Empty;
+
+using SomePrimitiveTaskClient = TaskClient<SomePrimitiveCommand, SomePrimitiveResult>;
+using SomePrimitiveTaskServer = TaskServer<SomePrimitiveCommand, SomePrimitiveResult>;
+
+template<>
+inline const char * getTaskName<SomePrimitiveCommand, SomePrimitiveResult>()
+{
+  return "SomePrimitiveTask";
+}
+
+}  // namespace nav2_tasks
+```
+
+For this example we arbitrarily pick `geometry_msgs::msg::Point` and `std_msgs::msg::Empty` as message types for command and result.
+
+Next we define the class for our new primitive. This class should derive from `MotionPrimitive` and use the command and result messages defined on the corresponding task.
+
+```cpp
+class SomePrimitive : public MotionPrimitive<nav2_tasks::SomePrimitiveCommand, nav2_tasks::SomePrimitiveResult>
+```
+
+On the implementation of `SomePrimitive` all we do is override `onRun` and `onCycleUpdate`.
+
+```cpp
+using nav2_tasks
+
+TaskStatus SomePrimitive::onRun(const SomePrimitiveCommand::SharedPtr command)
+{
+    /* onRun code */
+}
+
+TaskStatus SomePrimitive::onCycleUpdate(SomePrimitiveResult & result)
+{
+    /* onCycleUpdate code */
+}
+```
+
+The `onRun` method is the entry point for the primitive and here we should:
+- Catch the command.
+- Perform checks before the main execution loop.
+- Possibly do some initialization.
+- Return a `nav2_tasks::TaskStatus` given the initial checks.
+
+The `onCycleUpdate` method is called periodically until it returns `FAILED` or `SUCCEEDED`, here we should:
+- Set the robot in motion.
+- Perform some unit of work.
+- Check if the robot state, determine if work completed
+- Return a `nav2_tasks::TaskStatus`.
+
+### Executing a primitive
+
+Primitives use the `nav2_tasks` interface, so we need to define the task client:
+
+```cpp
+nav2_tasks::TaskClient<SomePrimitiveCommand, SomePrimitiveResult> some_primitive_task_client;
+```
+
+To send requests we create the command and sent it over the client:
+
+```cpp
+SomePrimitiveCommand::SharedPtr command;
+// Fill command
+some_primitive_task_client.sendCommand(command)
+```
+
+### (optional) Define the Behavior Tree action node
+
+For using motion primitivies within a behavior tree such as [bt_navigator](../nav2_bt_navigator/README.md##Navigation-Behavior-Trees), then a corresponding *action node* needs to be defined. Checkout `nav2_tasks` for examples on how to implement one.
+
+## Plans
+
+- Check for collision before executing a primitive. Issues [379](https://github.com/ros-planning/navigation2/issues/379) and [533](https://github.com/ros-planning/navigation2/issues/533).
+- Remove the stop primitive, move the funcionality to the robot class. Issue [575](https://github.com/ros-planning/navigation2/issues/575)
+- Consider moving `nav2_motion_primitives` altogether to the `nav2_robot` package. Issue [378](https://github.com/ros-planning/navigation2/issues/378).
+- Define primitives that describe arcs.
```

Comment:
Do you really think these extra primitives and primitive features are worth it? Everything the motion primitives do can also be done by using the rest of the nav stack.

Reply:
Or navigation could be done entirely by linking primitives :) 

Depending on the feedback from the community we might want to develop this package and support a wide variety of primitives to support all kinds of task, navigation, docking, etc.